### ISSUE TITLE
Suspicious honggfuzz_asan_harfbuzz issue

### ISSUE THREAD

**Issue by behdad** [google/oss-fuzz#7435]
Hello,

I'm writing about this issue https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=44212 (and at least one related issue found in the past). We've been scratching our head over it for a while and can't make sense of it. Any change to the code to debug it make it go away, and when we catch it in the debugger there's no debugging symbols.

Also, is it normal that such an access violation would happen when building with one engine (honggfuzz) but not another, with the same input?

Basically what it looks to us is that this line of code:
```
    if (unlikely (i >= len)) return Null (Type); 
```
when both `i` and `len` are `0`, is not working as expected, but we have no clue why. As I said, any changes to the code to facilitate debugging makes it work as expected.

Thanks for any insights you might have.

**Comment by jonathanmetzman** [comment]
>Any change to the code to debug it make it go away, and when we catch it in the debugger there's no debugging symbols.

Gnarly heisenbug.

I think it's likely this is a bug in honggfuzz since according to CF this doesn't reproduce on any other build.
@robertswiecki I haven't looked into this yet, but maybe you'll find it interesting.

**Comment by behdad** [comment]
afl has discovered this issue as well:
https://oss-fuzz.com/testcase-detail/4919078107611136

So it's definitely something on our side. It again, disappears as soon as I add a printf... But closing issue here. Thanks.

**Comment by behdad** [comment]
This was a real issue after all; had to do with compiler reordering reads where we were doing some unsafe magic. Solution was to insert memory barriers.
https://github.com/harfbuzz/harfbuzz/commit/f3151b6582a38fb1a377eb6070b8cecedb2ea711

**Comment by behdad** [comment]
> ```
>     if (unlikely (i >= len)) return Null (Type); 
> ```

It was batching a read before this line and a read after this line together.
